Tenants API

imtayadeway opened the pull request:
Resolves https://trello.com/c/QdPWdhWN

@abellotti @gtanzillo please review

imtayadeway commented:
@miq-bot add-label core api

imtayadeway commented:
@miq-bot add-label enhancement

imtayadeway commented:
@miq-bot add-label core, api

abellotti commented on config/api.yml:
```diff
@@ -859,6 +859,28 @@
+    :collection_actions:
+      :post:
+      - :name: create
+        :identifier: rbac_tenant_add
```
add the edit and delete in the collection post, supports multiple tenant edit/delete.

abellotti commented on config/api.yml:
```diff
@@ -859,6 +859,28 @@
+      :delete:
+      - :name: delete
+        :identifier: rbac_tenant_delete
+      :put:
```
no need to put section,  code leverages post/edit for entitlement.

abellotti commented on spec/requests/api/tenants_spec.rb:
```diff
@@ -0,0 +1,180 @@
+    end
+  end
+
+  context "with an appropriate role" do
```
without?

abellotti commented on spec/requests/api/tenants_spec.rb:
```diff
@@ -0,0 +1,180 @@
+
+      expect_request_success
+    end
+
```
maybe add a spec to test adding a tenant with an invalid parent ?

abellotti commented:
looks good @imtayadeway  minor comments, and the rubocop warn. Thanks.

imtayadeway commented:
@abellotti addressed your comments and the rubocop offense (and some new rubocop offenses) in subsequent commits. Anything to add?

One thing I meant to bring up yesterday - do the `GET`s need to have RBAC applied to them? I see that there are roles for lists and shows for tenants, but could not get this to work so left it out.

I'll squash the smaller stuff before this gets merged :)

imtayadeway commented:
@h-kataria can you advise on the proper roles for list/show of tenants in the API?

h-kataria commented:
@imtayadeway all the tenant features are listed in miq_product_features.yml, https://github.com/ManageIQ/manageiq/blob/master/db/fixtures/miq_product_features.yml#L2290 for list & show of Tenants you might need rbac_tenant_view feature that allows both rbac_tenant_show_list & rbac_tenant_show features. Let me know if you need more information.

imtayadeway commented:
Thanks @h-kataria, that helps.

Added some infrastructure here https://github.com/ManageIQ/manageiq/pull/4754 to support adding RBAC validation for the show/list actions

imtayadeway commented:
@abellotti just rebased to incorporate RBAC for GET requests, with added tests

imtayadeway commented:
@abellotti just applied your patch and waiting for green. Thanks!

abellotti commented:
:cake:
